MaxText/tests/attention_test.py

```diff
@@ -1,110 +1,402 @@
-    if self.BS > 1:
+      return [
+          jnp.arange(self.max_target_length, dtype=jnp.int32)
+          for _ in range(self.global_batch_size)
+      ]
+
+    if self.global_batch_size > 1:
       decoder_positions = jnp.stack(batch_positions())
-    decoder_mask = self.get_decoder_mask()
+
+    decoder_mask = self.get_decoder_mask(dtype)
     return lnx, decoder_mask, decoder_segment_ids, decoder_positions
 
-  def test_attention(self):
-    lnx, decoder_mask, decoder_segment_ids, decoder_positions = self.get_data()
-
-    mha_output = self.attention.apply(
-            self.variable,
-            lnx,
-            lnx,
-            decoder_segment_ids = decoder_segment_ids,
-            attention_type='mha',
-            inputs_positions = decoder_positions,
-            mask = decoder_mask,
-            bias = None,
-            deterministic=True,
-            decode=False,
-            rngs={'aqt': self.rng}
+  def test_flash_multihead_attention(self):
+    """Test Flash MHA."""
+
+    mha_attention_layer = MultiHeadDotProductAttention(
+        num_heads=self.num_heads,
+        head_dim=self.head_dim,
+        mesh=self.mesh,
+        dtype=self.dtype,
+        dropout_rate=self.cfg.dropout_rate,
+        name='self_attention',
+    )
+    variable = mha_attention_layer.init(
+        {'params': self.rng, 'aqt': self.rng},
+        jnp.ones(
+            (self.global_batch_size, self.max_target_length, self.embed_dim)),
+        jnp.ones(
+            (self.global_batch_size, self.max_target_length, self.embed_dim)),
+        'mha',
+    )
+
+    flash_attention_layer = FlashMultiHeadDotProductAttention(
+        num_heads=self.num_heads,
+        head_dim=self.head_dim,
+        mesh=self.mesh,
+        dtype=self.dtype,
+        dropout_rate=self.cfg.dropout_rate,
+        name='self_attention',
+        max_target_length=self.max_target_length
+    )
+
+    lnx, decoder_mask, decoder_segment_ids, decoder_positions = self.get_data(
+        self.dtype)
+    bias = None
+    mha_output = mha_attention_layer.apply(
+        variable,
+        lnx,
+        lnx,
+        decoder_segment_ids=decoder_segment_ids,
+        inputs_positions=decoder_positions,
+        mask=decoder_mask,
+        bias=bias,
+        deterministic=True,
+        decode=False,
+        rngs={'aqt': self.rng},
+    )
+    flash_output = flash_attention_layer.apply(
+        variable,
+        lnx,
+        lnx,
+        decoder_segment_ids=decoder_segment_ids,
+        inputs_positions=decoder_positions,
+        mask=decoder_mask,
+        bias=bias,
+        deterministic=True,
+        decode=False,
+        rngs={'aqt': self.rng},
+    )
+
+    self.assertTrue(
+        jax.numpy.allclose(
+            flash_output, mha_output, rtol=1e-01, atol=1e-01, equal_nan=False
+        )
+    )
+
+  def test_multihead_attention(self):
+    """Test MultiHeadDotProductAttention."""
```

Comment:
I find the reimplementation kind of weird! Shouldn't we just test that MHA == FLASH_MHA?

Reply:
Then how we know MHA or Flash MHA is correct?

Reply:
Who is our ground truth then?